### PR TITLE
AmoebaNet Google Cloud TPU - Tensorflow 1.8 - Imagenet training

### PR DESCRIPTION
**Pull request by bignamehyp**
Hi,
We train a new Amoeba Net model on Google Cloud TPU and this is the result. Please have a review.

Ref:
https://arxiv.org/abs/1802.01548

Thanks
Yanping

### CODE REVIEW

**Comment by bignamehyp** [comment]
Gentle ping. We have to publish this result by noon April 17.  Can you please take a look? Thank you very much. @codyaustun @deepakn94

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  "version": "v1.0",
+  "author": "Google Cloud TPU",
+  "authorEmail": "cloud-tpu-dawnbench@googlegroups.com",
+  "framework": "TensorFlow v1.7rc1",
+  "codeURL": "https://github.com/tensorflow/tpu/tree/master/models/experimental/amoeba_net",
+  "model": "AmoebaNet-D N6F256",
+  "hardware": "GCP n1-standard-2, Cloud TPU",
+  "costPerHour": 6.595,
+  "timestamp": "2018-04-13",
+  "misc": {
```
Comment (codyaustun) [inline ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: Because this is a new architecture, please add a link to the paper under `misc`.

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    "optimizer": "RMSProp",
+    "Moving Average Decay": 0.9999,
```
Comment (codyaustun) [inline ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: To be consistent, I would rename this field `movingAverageDecay`

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    "weightDecay": 0.00004,
+    "batchSize": 1024,
+    "warmUp": "the learning rate starts at 0, then it increases linearly per iteration to the inital learning rate for the first 1% of the total steps",
+    "exponentiial decay": "The initial learning rate is 2.56, it decrease by a factor of 0.88 every one epoch"
```
Comment (codyaustun) [inline ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: It looks like there is a typo in the field name. I think you want `exponentialDecay`.

Also, can you add another field for the number of epochs? From your README and TSV, I believe you ran for 35 epochs, but it would be good to make it clear here.

**Review by codyaustun — CHANGES_REQUESTED** [review]
@bignamehyp I took a quick look at the JSON and TSV files to give you some feedback. I'll look at the paper and code more closely this evening and follow up.

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+  "version": "v1.0",
+  "author": "Google Cloud TPU",
+  "authorEmail": "cloud-tpu-dawnbench@googlegroups.com",
+  "framework": "TensorFlow v1.7rc1",
+  "codeURL": "https://github.com/tensorflow/tpu/tree/master/models/experimental/amoeba_net",
```
Comment (codyaustun) [inline ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: Can you either change this to a specific commit hash or include the commit hash in `misc`?

**Comment by bignamehyp** [comment]
Thank you very much for your reviews. Made changes based on your comments.

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json (hunk 4 above)
Comment (bignamehyp) [reply ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: Done

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json (hunk 3 above)
Comment (bignamehyp) [reply ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: Done

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json (hunk 2 above)
Comment (bignamehyp) [reply ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: Done

File: ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json (hunk 1 above)
Comment (bignamehyp) [reply ImageNet/train/google_amoeba_net_d_tpu_tensorflow17.json]: Done

**Review by codyaustun — APPROVED** [review]

**Comment by codyaustun** [comment]
The code looks good to me although it is a bit complicated. AmoebaNet-D isn't described in the paper. From [model_specs.py#L44](https://github.com/tensorflow/tpu/blob/1acd2c4d48013ea1623ddbf9f28166a85404213a/models/experimental/amoeba_net/model_specs.py#L44) and [L79](https://github.com/tensorflow/tpu/blob/1acd2c4d48013ea1623ddbf9f28166a85404213a/models/experimental/amoeba_net/model_specs.py#L79), I see how the normal and reduction cell are defined through [network_utils.py#L324](https://github.com/tensorflow/tpu/blob/1acd2c4d48013ea1623ddbf9f28166a85404213a/models/experimental/amoeba_net/network_utils.py#L324), but it is pretty painful. I had to draw it out with a pencil and paper. Can you include a diagram in case someone wants to reimplement this model?

**Comment by codyaustun** [comment]
Also, can you briefly describe how you created this learning rate schedule?

**Comment by bignamehyp** [comment]
Certainly, we will add a link to the diagram in the comments of network_utils. The network architecture is evolved so it's somehow difficult to understand and visualize.  

This learning rate schedule is basically an exponential decay one with standard warmup.  The learning rate decay factor is computed so that the final learning rate at epoch 35 is about 1% of the initial one.

**Comment by codyaustun** [comment]
Great! Something like Figure S1 for AmoebaNet-D would be perfect.

So the hyperparameters for the exponential decay were just found by searching right?

**Comment by bignamehyp** [comment]
@codyaustun, thank you very much for reviewing in the evening.

 I included links to cell diagrams in the misc field.  Please take another look. We will also include those diagrams in our code comment.
 
Actually, we only searched the neural architecture. And the learning rate schedule is just from empirical experience. There're too many free parameters to search for an exponential decay schedule with warm-up. We have been using exponential learning rate for the amoeba net family. And we knew resent 50 decreased the learning rate by a factor of 100 after 35 epochs so we used it to compute our learning rate decay factor 0.88^35 = 0.01+. We did a grid search over five possible initial learning rates and choose the best one with highest validation accuracy.  It's not necessary the best learning rate schedule for this model.

**Comment by bignamehyp** [comment]
If possible, can you please merge this PR and publish the results on the website by noon today (April/17/2018)? We have a deadline by then. I really appreciated your quick review and prompt response!

**Comment by codyaustun** [comment]
@bignamehyp I just noticed an error updating the website. The `top1Accuracy` and `top5Accuracy` columns should be multiplied by 100. I am sorry I didn't catch this during the review.  Can you submit a new pull request updating the column values? The website won't update the leaderboard until this is changed.

**Comment by bignamehyp** [comment]
@codyaustun  thank you very much for merging #13. No worries. I created #14 to fix it.
